iam-token-manager/v1.ts

```diff
@@ -0,0 +1,198 @@
+ * limitations under the License.
+ */
+
+import extend = require('extend');
+import { sendRequest } from '../lib/requestwrapper';
+
+export type Options = {
+  iam_apikey?: string;
+  iam_access_token?: string;
+  iam_url?: string;
+}
+
+export interface IamTokenData {
+  access_token: string;
+  refresh_token: string;
+  token_type: string;
+  expires_in: number;
+  expiration: number;
+}
+
+export class IamTokenManagerV1 {
+  name: string;
+  serviceVersion: string;
+  protected url: string;
+  protected tokenInfo: IamTokenData;
+  private iam_apikey: string; // tslint:disable-line variable-name
+  private userAccessToken: string;
+
+  /**
+   * IAM Token Manager Service
+   *
+   * Retreives, stores, and refreshes IAM tokens.
+   *
+   * @param {Object} options
+   * @param {String} options.iam_apikey
+   * @param {String} options.iam_access_token
+   * @param {String} options.iam_url - url of the iam api to retrieve tokens from
+   * @constructor
+   */
+  constructor(options: Options) {
+    this.url = options.iam_url || 'https://iam.ng.bluemix.net/identity/token';
+    this.tokenInfo = {} as IamTokenData;
+    if (options.iam_apikey) {
+      this.iam_apikey = options.iam_apikey;
+    }
+    if (options.iam_access_token) {
+      this.userAccessToken = options.iam_access_token;
+    }
+  }
+
+  /**
+   * This function sends an access token back through a callback. The source of the token
+   * is determined by the following logic:
+   * 1. If user provides their own managed access token, assume it is valid and send it
+   * 2. If this class is managing tokens and does not yet have one, make a request for one
+   * 3. If this class is managing tokens and the token has expired, refresh it
+   * 4. If this class is managing tokens and has a valid token stored, send it
+   *
+   * @param {Function} cb - callback function that the token will be passed to
+   */
+  public getToken(cb: Function) {
+    if (this.userAccessToken) {
+      // 1. use user-managed token
+      return cb(null, this.userAccessToken);
+    } else if (!this.tokenInfo.access_token) {
+      // 2. request an initial token
+      this.requestToken(null, (err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else if (this.isTokenExpired()) {
+      // 3. refresh a token
+      this.refreshToken(null, (err, tokenResponse) => {
+        this.saveTokenInfo(tokenResponse);
+        return cb(err, this.tokenInfo.access_token);
+      });
+    } else {
+      // 4. use valid managed token
+      return cb(null, this.tokenInfo.access_token);
+    }
+  }
+
+  /**
+   * Request an IAM token using an API key.
+   *
+   * @param {string} authHeader - The authorization header to send to the service. Defaults to 'Basic Yng6Yng='.
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  public requestToken(authHeader: string, cb: Function): void {
```

Comment:
async calls should always have `params` and `callback`. Synchronous functions can have positional parameters because there will not be a `callback` that always needs to be there so it can be flexible.